Frameworks/CoreGraphics/CGPattern.mm

```diff
@@ -17,160 +17,78 @@
-        colorSpace = CGColorSpaceCreateDeviceRGB();
-    } else {
-        bitmapInfo = kCGImageAlphaNone;
-        colorSpace = CGColorSpaceCreateDeviceGray();
+    inline const CGRect Bounds() const {
+        return _bounds;
     }
 
-    patternCtx = CGBitmapContextCreate(NULL,
-                                       ceilf(outWidth),
-                                       ceilf(outHeight),
-                                       8,
-                                       4 * ceilf(outWidth),
-                                       CGColorSpaceCreateDeviceRGB(),
-                                       kCGImageAlphaPremultipliedFirst | kCGBitmapByteOrder32Big);
-    CGRect rect = { 0, 0, width, height };
-    CGContextDrawTiledImage(patternCtx, rect, tilePattern);
-
-    CGColorSpaceRelease(colorSpace);
-    CGImageRelease(tilePattern);
-
-    if (generatedImage) {
-        CGImageRelease(generatedImage);
+    inline void IssuePatternCallback(CGContextRef context) const {
+        _callbacks.drawPattern(_info, context);
     }
 
-    // TODO: Caching.
-    generatedImage = CGBitmapContextCreateImage(patternCtx);
-    CGImageRetain(generatedImage);
-
-    CGContextRelease(patternCtx);
-
-    return generatedImage;
-}
-
-- (CGAffineTransform)getPatternTransform {
-    return matrix;
-}
-
-- (void)dealloc {
-    if (generatedImage) {
-        CGImageRelease(generatedImage);
+    inline CGAffineTransform GetTransformMatrix() const {
+        return _transformMatrix;
     }
-    if (simpleImage) {
-        CGImageRelease(simpleImage);
+
+    inline CGRect FinalPatternSize() const {
+        return { CGPointZero, { _xStep, _yStep } };
     }
-    if (callbacks.releaseInfo) {
-        callbacks.releaseInfo(info);
+
+    inline bool Colored() const {
+        return _isColored;
     }
-    [super dealloc];
-}
-@end
-
-CGPatternRef CGPatternCreateColorspace(void* info,
-                                       CGRect bounds,
-                                       CGAffineTransform matrix,
-                                       CGFloat xStep,
-                                       CGFloat yStep,
-                                       CGPatternTiling tiling,
-                                       bool isColored,
-                                       const CGPatternCallbacks* callbacks,
-                                       __CGSurfaceFormat fmt) {
-    CGPattern* ret = [CGPattern new];
-
-    // TODO: Obey the colorspace.
-    ret->info = info;
-    ret->bounds = bounds;
-    ret->matrix = matrix;
-    ret->xStep = xStep;
-    ret->yStep = yStep;
-    ret->tiling = tiling;
-    ret->isColored = isColored;
-    ret->callbacks = *callbacks;
-    ret->surfaceFmt = fmt;
-
-    return ret;
-}
+
+private:
+    woc::StrongCF<CGImageRef> _generatedImage;
+    woc::StrongCF<CGImageRef> _simpleImage;
+    void* _info;
+    CGRect _bounds;
+    CGAffineTransform _transformMatrix;
+    CGFloat _xStep;
```

Comment:
it's much more implicit this way.

---
In reply to: [109041128](https://github.com/Microsoft/WinObjC/pull/2374#discussion_r109041128) [](ancestors = 109041128)